src/components/SigningModals/TransactionStepper/TransactionStepper.js

```diff
@@ -0,0 +1,247 @@
+  STEP_ERROR,
+  STEP_PROMPTING,
+  STEP_SUCCESS,
+  STEP_WAITING,
+  STEP_WORKING,
+} from './stepper-statuses'
+import useStepperLayout from './useStepperLayout'
+import Step from './Step/Step'
+
+const AnimatedDiv = animated.div
+
+function initialStepState(steps) {
+  return steps.map((_, i) => {
+    return {
+      status: i === 0 ? STEP_PROMPTING : STEP_WAITING,
+      hash: null,
+    }
+  })
+}
+
+function reduceSteps(steps, [action, stepIndex, value]) {
+  if (action === 'setHash') {
+    steps[stepIndex].hash = value
+    return [...steps]
+  }
+  if (action === 'setStatus') {
+    steps[stepIndex].status = value
+    return [...steps]
+  }
+  return steps
+}
+
+function TransactionStepper({ steps, onComplete, className }) {
+  const theme = useTheme()
+  const [stepperStage, setStepperStage] = useState(0)
+  const [stepState, updateStep] = useReducer(
+    reduceSteps,
+    initialStepState(steps)
+  )
+  const [firstStart, setFirstStart] = useState(true)
+
+  const {
+    outerBoundsRef,
+    innerBoundsRef,
+    layout,
+    measuring,
+  } = useStepperLayout()
+
+  const canPerformUpdate = useRef(true)
+
+  const stepsCount = steps.length - 1
+
+  const onStart = useCallback(() => {
+    // Don’t animate on first render
+    if (firstStart) {
+      setFirstStart(false)
+    }
+  }, [firstStart])
+
+  const renderStep = (stepIndex, showDivider) => {
+    const title = steps[stepIndex][0]
+    const { status, hash } = stepState[stepIndex]
+
+    return (
+      <li
+        key={stepIndex}
+        css={`
+          display: flex;
+        `}
+      >
+        <Step
+          title={title}
+          number={stepIndex + 1}
+          status={status}
+          showDivider={showDivider}
+          transactionHash={hash}
+        />
+      </li>
+    )
+  }
+
+  const renderSteps = () => {
+    return steps.map((_, index) => {
+      const showDivider = index < stepsCount
+
+      return renderStep(index, showDivider)
+    })
+  }
+
+  const updateStepStatus = useCallback(
+    status => {
+      if (canPerformUpdate.current) {
+        updateStep(['setStatus', stepperStage, status])
+      }
+    },
+    [stepperStage]
+  )
+
+  const updateHash = useCallback(
+    hash => {
+      if (canPerformUpdate.current) {
+        updateStep(['setHash', stepperStage, hash])
+      }
+    },
+    [stepperStage]
+  )
+
+  const handleSign = useCallback(() => {
+    const signProcess = steps[stepperStage][1]
+
+    // Always start new step in the "Prompting" state
+    updateStepStatus(STEP_PROMPTING)
+
+    // Pass state updates as render props to signProcess
+    signProcess({
+      setStepHash: hash => updateHash(hash),
+      setStepWorking: () => updateStepStatus(STEP_WORKING),
+      setStepError: () => updateStepStatus(STEP_ERROR),
+      setStepSuccess: () => {
+        updateStepStatus(STEP_SUCCESS)
+
+        // Advance to next step or fire complete callback
+        if (stepperStage !== stepsCount && canPerformUpdate.current) {
```

Comment:
Small nitpick, we can lift the `canPerformUpdate` check to be above both of the other checks, so it's more obvious they're exclusive (i.e. by using `if else`)